extensions/vscode/src/multiStepInputs/newCredential.ts

```diff
@@ -174,96 +214,159 @@ export async function newCredential(
+          return Promise.resolve(undefined);
+        },
+        finalValidation: async (input: string) => {
+          // validate that the API key is formed correctly
+          const errorMsg = checkSyntaxApiKey(input);
+          if (errorMsg) {
             return Promise.resolve({
-              message: `Error: Invalid API Key (${testResult.data.error.msg}).`,
+              message: `Error: Invalid API Key (${errorMsg}).`,
               severity: InputBoxValidationSeverity.Error,
             });
           }
-          // we have success, but credentials.test may have returned a different
-          // url for us to use.
-          if (testResult.data.url) {
-            validatedURL = testResult.data.url;
+          // url should always be defined by the time we get to this step
+          // but we have to type guard it for the API
+          const serverUrl =
+            typeof state.data.url === "string" ? state.data.url : "";
+          try {
+            const testResult = await api.credentials.test(
+              serverUrl,
+              !extensionSettings.verifyCertificates(), // insecure = !verifyCertificates
+              input,
+            );
+            if (testResult.status !== 200) {
+              return Promise.resolve({
+                message: `Error: Invalid API Key (unable to validate API Key - API Call result: ${testResult.status} - ${testResult.statusText}).`,
+                severity: InputBoxValidationSeverity.Error,
+              });
+            }
+            if (testResult.data.error) {
+              return Promise.resolve({
+                message: `Error: Invalid API Key (${testResult.data.error.msg}).`,
+                severity: InputBoxValidationSeverity.Error,
+              });
+            }
+            // we have success, but credentials.test may have returned a different
+            // url for us to use.
+            if (testResult.data.url) {
+              validatedURL = testResult.data.url;
+            }
+          } catch (e) {
+            return Promise.resolve({
+              message: `Error: Invalid API Key (${getMessageFromError(e)})`,
+              severity: InputBoxValidationSeverity.Error,
+            });
           }
-        } catch (e) {
-          return Promise.resolve({
-            message: `Error: Invalid API Key (${getMessageFromError(e)})`,
-            severity: InputBoxValidationSeverity.Error,
-          });
+          return Promise.resolve(undefined);
+        },
+        shouldResume: () => Promise.resolve(false),
+        ignoreFocusOut: true,
+      });
+
+      // only one of api key and snowflake connection should be configured
+      state.data.apiKey = apiKey;
+      state.data.snowflakeConnection = "";
+      state.data.url = validatedURL;
+      state.lastStep = thisStepNumber;
+      return (input: MultiStepInput) => inputSnowflakeConnection(input, state);
+    }
+    return inputSnowflakeConnection(input, state);
```

Comment:
This is a bit odd, calling `inputSnowflakeConnection` regardless of the serverType check. With the suggestion above this method can stay as it is.

Reply:
In the existing pattern, each step needs to implement the logic to suppress itself, and shouldn't rely upon previous steps to determine it.